File: leetcode/binary_search/2323_Find_Minimum_Time_to_Finish_All_Jobs_II/codes/bs.c
#include <stdio.h>
#include <stdlib.h>

int cmpAsc(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}

int canFinish(int* jobs, int* workers, int n, int days) {
    int i = 0, j = 0;
    while (i < n && j < n) {
        if ((long long)workers[j] * days >= jobs[i]) i++;
        j++;
    }
    return i == n;
}

int minimumTime(int* jobs, int jobsSize, int* workers, int workersSize) {
    qsort(jobs, jobsSize, sizeof(int), cmpAsc);
    qsort(workers, workersSize, sizeof(int), cmpAsc);

    int left = 1, right = 100000; 
    for (int i = 0; i < jobsSize; i++)
        if (jobs[i] > right) right = jobs[i];

    while (left < right) {
        int mid = (left + right) / 2;
        if (canFinish(jobs, workers, jobsSize, mid))
            right = mid;
        else
            left = mid + 1;
    }
    return left;
}

File: leetcode/binary_search/2323_Find_Minimum_Time_to_Finish_All_Jobs_II/codes/bs.py
from typing import List

class Solution:
    def minimumTime(self, jobs: List[int], workers: List[int]) -> int:
        jobs.sort()
        workers.sort()

        def canFinish(days: int) -> bool:
            i = 0  
            j = 0  
            n = len(jobs)

            while i < n and j < n:
                if workers[j] * days >= jobs[i]:
                    i += 1 
                j += 1 
            return i == n 

        left, right = 1, max(jobs)
        while left < right:
            mid = (left + right) // 2
            if canFinish(mid):
                right = mid
            else:
                left = mid + 1
        return left
